### PR TITLE
[Bridge] Polygon Edge Bridge

### PR DESCRIPTION
**Pull request by Kourin1996**
# Description

~~This PR adds Bridge which assembles bridge module and manage Tracker and SAM Pool.~~
~~This PR is based on https://github.com/0xPolygon/polygon-edge/pull/386, https://github.com/0xPolygon/polygon-edge/pull/412~~
~~Please don't push the change of Tracker and SAM Pool to this branch directly.~~

This is base branch for bridge stuffs. Currently this PR has the function for bridge v1

To start server with Bridge, please add following parameters.
Please note these parameters for debug, the names may change.

```
$ ./polygon-edge --bridge --bridge-rootchain-url ws://127.0.0.1:11002/ws --bridge-rootchain-contract 0xA82373a89b8c7c08ad6c24d4961Df43179B690e2 --bridge-rootchain-confirmations 5
```

Please provide a detailed description of what was done in this PR

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [ ] I have assigned this PR to myself
- [ ] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [ ] I have tested this code with the official test suite
- [ ] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

Fixes EDGE-402 and EDGE-396

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -577,6 +610,17 @@ func (s *Server) Close() {
 		}
 	}
 
+	// if s.tracker != nil {
+	// 	if err := s.tracker.Stop(); err != nil {
+	// 		s.logger.Error("failed to shutdown tracker", "err", err)
+	// 	}
+	// }
```
Comment (lazartravica) [inline server/server.go]: Can we remove this comment?

File: server/server.go [hunk 2]
```diff
@@ -180,6 +185,26 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
 		m.txpool.SetSigner(signer)
 	}
 
+	// bridge
+	if config.Bridge.Enable {
+		// FIXME: use new key?
```
Comment (lazartravica) [inline server/server.go]: No need to, reusing the validator key for this purpose is intended.

Do you see any reason to use a new key?

File: consensus/ibft/ibft.go [hunk 3]
```diff
@@ -746,6 +750,15 @@ func (i *Ibft) runAcceptState() { // start new round
 		return
 	}
 
+	if i.bridge != nil {
+		i.bridge.SetValidators(snap.Set, uint64(snap.Set.Len()))
+
+		for _, msg := range i.bridge.GetReadyMessages() {
+			fmt.Printf("ReadyMessage height=%d, hash=%+v, body=%+v, signatures=%d\n", parent.Number+1, msg.Hash, msg.Body, len(msg.Signatures))
```
Comment (lazartravica) [inline consensus/ibft/ibft.go]: Please fix the linter error by splitting argument into mutiple lines

File: command/helper/helper.go [hunk 4]
```diff
@@ -446,6 +447,7 @@ func ReadConfig(baseCommand string, args []string) (*Config, error) {
 
 	flags.StringVar(&cliConfig.LogLevel, "log-level", "", "")
 	flags.BoolVar(&cliConfig.Seal, "seal", false, "")
+	flags.BoolVar(&cliConfig.Tracker, "tracker", false, "")
```
Comment (lazartravica) [inline command/helper/helper.go]: Why do we need separate configuration parameter for tracker and for bridge? Doesn't enabling the bridge enable the tracker as well?

File: command/helper/config.go [hunk 5]
```diff
@@ -107,6 +124,7 @@ func (c *Config) BuildConfig() (*server.Config, error) {
 
 	conf.Chain = cc
 	conf.Seal = c.Seal
+	// conf.Tracker = c.Tracker
```
Comment (lazartravica) [inline command/helper/config.go]: We can remove this comment

File: bridge/tracker/tracker_test.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+// nolint
+func Test_InvalidWSResponse(t *testing.T) {
+	testCases := []struct {
+		name     string
+		response []byte
+	}{
+		{
+			"invalid response",
+			[]byte(`{
```
Comment (lazartravica) [inline bridge/tracker/tracker_test.go:62]: whoever wrote the invalid block example made me laugh, feel free to resolve this comment :)

File: bridge/bridge.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+		)
+
+		return
+	}
+
+	b.sampool.AddSignature(&sam.MessageSignature{
+		Hash:      message.Hash,
+		Address:   sender,
+		Signature: message.Signature,
+	})
```
Comment (lazartravica) [inline bridge/bridge.go]: This doesn't seem good, we shouldn't blindly sign any message we receive from a validator and put it into SAM pool. We should only put it there if we see it first.

File: bridge/tracker/rootchain.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+	"os"
+	"path/filepath"
+)
+
+const (
+	//	required block depth for fetching events on the rootchain
+	DefaultBlockConfirmations = 6
+
+	//	Ropsten testnet
+	//rootchainWS   = "wss://ropsten.infura.io/ws/v3/17eac086ff36442ebd43737400eb71ca"
```
Comment (lazartravica) [inline bridge/tracker/rootchain.go]: please remove

File: bridge/tracker/rootchain.go [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+//	Number       string        `json:"number"`
+//	ParentHash   types.Hash    `json:"parentHash"`
+//	ReceiptsRoot types.Hash    `json:"receiptsRoot"`
+//	Sha3Uncles   types.Hash    `json:"sha3Uncles"`
+//	StateRoot    types.Hash    `json:"stateRoot"`
+//	Timestamp    string        `json:"timestamp"`
+//	TxRoot       types.Hash    `json:"transactionsRoot"`
+//	MixHash      types.Hash    `json:"mixHash"`
+//	Hash         types.Hash    `json:"hash"`
+//}
```
Comment (lazartravica) [inline bridge/tracker/rootchain.go:106]: @dbrajovic why do we need this commented out struct? i can't remember what was the conclusion to the things around this? on which occasions does the first vs the second structs work? were we able to notice the difference in a textual diff??

File: bridge/tracker/rootchain.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+//	handleWSResponse parses the json response
+//	received by the websocket into a header struct.
+func (s *subscription) handleWSResponse(response []byte) {
+	//	parse ws response
+	header := &ethHeader{}
+	if err := json.Unmarshal(response, header); err != nil {
+		s.errorCh <- fmt.Errorf("unable to parse header - err: %w", err)
+		return
```
Comment (lazartravica) [inline bridge/tracker/rootchain.go:180]: @dbrajovic pls fix linter error

File: bridge/tracker/tracker.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+
+	//	notify each matched log event
+	t.notify(logs...)
+}
+
+//	calculateRange determines the next range of blocks
+//	to query for events.
+func (t *Tracker) calculateRange(header *ethHeader) (from, to *big.Int) {
+	//	extract block number from header field
+	//	TODO: polygon-edge header
```
Comment (lazartravica) [inline bridge/tracker/tracker.go:213]: @dbrajovic what do these TODOs mean?

**Review by lazartravica — CHANGES_REQUESTED** [review]
Beautifully done!

I have found a potential issue in the addRemote message bridge method, have some comments around the tracker

File: bridge/bridge.go (hunk 7 above)
Comment (Kourin1996) [reply bridge/bridge.go]: I don't agree with that. Is it always true that the RootChain event comes first and signature from other validators reach after this?

Comment (Kourin1996) [reply bridge/bridge.go]: @lazartravica

File: command/helper/helper.go (hunk 4 above)
Comment (Kourin1996) [reply command/helper/helper.go]: Removed at b9bf70f4367a32c25825acd09c3526c486faeaff

File: command/helper/config.go (hunk 5 above)
Comment (Kourin1996) [reply command/helper/config.go]: Removed at b9bf70f4367a32c25825acd09c3526c486faeaff

File: server/server.go (hunk 1 above)
Comment (Kourin1996) [reply server/server.go]: Removed at eb8b4ccc7699cfb6e69ae0e1fdc027a0e8fc10fc

File: server/server.go (hunk 2 above)
Comment (Kourin1996) [reply server/server.go]: Removed comment at eb8b4ccc7699cfb6e69ae0e1fdc027a0e8fc10fc

File: consensus/ibft/ibft.go (hunk 3 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Fixed at e46c48c0f82afeaccc5b02abbfa5f09a54c9b94e

File: bridge/tracker/rootchain.go (hunk 9 above)
Comment (dbrajovic) [reply bridge/tracker/rootchain.go:106]: Conclusion was that if you wanna use polygon-edge as rootchain (as we did for PoC) then the `ethHeader` needs to be different - our header field types are different. I don't think we reached any consensus on this.

File: bridge/tracker/tracker.go (hunk 11 above)
Comment (dbrajovic) [reply bridge/tracker/tracker.go:213]: These TODOs are related to the ethereum vs polygon-edge `header` type. (https://github.com/0xPolygon/polygon-edge/pull/419#discussion_r818495762)

File: bridge/tracker/rootchain.go (hunk 8 above)
Comment (Kourin1996) [reply bridge/tracker/rootchain.go]: @dbrajovic

File: bridge/bridge.go (hunk 7 above)
Comment (lazartravica) [reply bridge/bridge.go]: @Kourin1996 You're absolutely right, thanks for the clarification 🎉

File: bridge/tracker/rootchain.go (hunk 8 above)
Comment (dbrajovic) [reply bridge/tracker/rootchain.go]: [Removed](https://github.com/0xPolygon/polygon-edge/pull/419/commits/217410b592f10aafe8d3ee8542653812a4f7ae8e)

File: bridge/tracker/rootchain.go (hunk 10 above)
Comment (dbrajovic) [reply bridge/tracker/rootchain.go:180]: [Fixed](https://github.com/0xPolygon/polygon-edge/pull/419/commits/217410b592f10aafe8d3ee8542653812a4f7ae8e)

**Comment by zivkovicmilos** [comment]
Closed due to archiving

**Pull request closed without merging**